working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+| JavaScript | Memory model is documented in [ECMA-262 section 13.0 *Memory Model*](https://262.ecma-international.org/13.0/#sec-memory-model)*.* This memory model is fairly straightforward: it guarantees sequential consistency for atomic operations, while leaving other operations unordered. |
+| Java       | Given in [Java Language Specification (JLS) section 17.4](https://docs.oracle.com/javase/specs/jls/se19/html/jls-17.html#jls-17.4) |
+| Kotlin     | Not documented. Kotlin/JVM effectively inherits Java's memory model. Kotlin/Native - does not have a specified memory model, but likely follows JVM's one as well. |
+| C++        | Given in the section [Multi-threaded executions and data races](https://eel.is/c++draft/intro.multithread) of the standard (since C++11). Notably very fine grained |
+| Rust       | No official memory model according to [reference](https://doc.rust-lang.org/reference/memory-model.html), however it is documented to "[blatantly inherit C++20 memory model](https://doc.rust-lang.org/nomicon/atomics.html)" |
+| Swift      | Defined to be consistent with C/C++. See [SE-0282](https://github.com/apple/swift-evolution/blob/main/proposals/0282-atomics.md) |
+| Go         | Given [here](https://go.dev/ref/mem): race free programs have sequential consistency, programs with races still have some non-deterministic but well-defined behavior. |
+
+## Platform capabilities
+
+When expanding Dart's capabilities we need to consider if this semantic can be implemented across the platforms that Dart runs on.
+
+### Native
+
+No blockers to implement any multithreading behavior. VM already has a concept of _isolate groups_: multiple isolates concurrently sharing the same heap and runtime infrastructure (GC, program structure, debugger, etc). 
+
+### Web (JS and Wasm)
+
+No shared memory multithreading currently (beyond unstructured binary data shared via `SharedArrayBuffer`). However there is a Strage 1 TC-39 proposal [JavaScript Structs: Fixed Layout Objects and Some Synchronization Primitives](https://github.com/tc39/proposal-structs) which introduces the concept of _struct_ - fixed shape mutable object which can be shared between different workers. Structs are very similar to `Shareable` objects we propose, however they can't have any methods associated with them. This makes structs unsuitable for representing arbitrary Dart classes - which usually have methods associated with them.
+
+Wasm GC does not yet have a similar concurrency story - but it seems reasonable to expect that eventually it will get something similar to structs.
+
+## `dart:*` race safety
+
+When implementing `dart:*` libraries we should keep racy access in mind. Consider for example the following code:
+
+```dart 
+class _GrowableList<T> implements List<T> {
+  int length;
+  final _Array storage;
+  
+  T operator[](int index) {
```

Comment:
Do we want to introduce a synchronized method to aid developers in writing atomic blocks of code? (Java style)

```dart 
class _GrowableList<T> implements List<T> {
  @shared
  static Lock lock;

  int length;
  final _Array storage;
  
  T operator[](int index) {
    synchronized(lock) {
      RangeError.checkValidIndex(index, this, "index", this.length);    
      return unsafeCast(storage[index]);
    }
  }
  
  T removeLast() {
    synchronized(lock) {
      final result = this[length];
      storage[length--] = null;
      return result;
    }
  } 
}
```

Or even a short-hand syntax:

```dart 
class _GrowableList<T> implements List<T> {
  @shared
  static Lock lock;

  int length;
  final _Array storage;
  
  T operator[](int index) synchronized {
    RangeError.checkValidIndex(index, this, "index", this.length);    
    return unsafeCast(storage[index]);
  }
  
  T removeLast() synchronized {
    final result = this[length];
    storage[length--] = null;
    return result;
  } 
}
```

Reply:
Allowing mutex-guarded methods is an option.
Doing it on `_GrowableList` is probably not a good idea.

Java had `StringBuffer`, which was synchronized on every member so it could safely be used concurrently.
That was too slow and [even confusing](https://stackoverflow.com/questions/16653119/imagine-a-real-concurrent-scenario-where-stringbuffer-should-be-used-than-string), and they introduced `StringBuilder` which wasn't synchronized, so that users could be efficient, and then do their own synchronization at the level where it made sense.

Making `_GrowableList` do thread synchronization on every access is very, very risky from a performance perspective. If you can do it with zero overhead for unshared lists, then sure.
If not, it's better to let users do their own synchronization.

Reply:
I think we should not do it. Explicit locking is better.